<?php
declare(strict_types=1);

namespace Tests;

use Bigcommerce\ORM\Entities\Customer;
use Bigcommerce\ORM\EntityManager;
use Bigcommerce\ORM\Mapper;
use Bigcommerce\ORM\Repository;

class RepositoryTest extends BaseTestCase
{
    /** @var \Bigcommerce\ORM\Repository */
    protected $repository;

    /** @var \Bigcommerce\ORM\EntityManager|\Prophecy\Prophecy\ProphecySubjectInterface */
    protected $entityManager;

    protected function setUp(): void
    {
        parent::setUp(); // TODO: Change the autogenerated stub
        $this->entityManager = $this->getEntityManager();
        $this->repository = new Repository($this->entityManager);
        $this->repository->setClassName(Customer::class);
    }

    /**
     * @covers \Bigcommerce\ORM\Repository::setEntityManager
     * @covers \Bigcommerce\ORM\Repository::setClassName
     * @covers \Bigcommerce\ORM\Repository::getEntityManager
     * @covers \Bigcommerce\ORM\Repository::getClassName
     */
    public function testSettersAndGetters()
    {
        $this->repository
            ->setClassName(Customer::class)
            ->setEntityManager($this->entityManager);

        $this->assertEquals(Customer::class, $this->repository->getClassName());
        $this->assertEquals($this->entityManager, $this->repository->getEntityManager());
    }

    /**
     * @covers \Bigcommerce\ORM\Repository::count
     * @throws \Bigcommerce\ORM\Client\Exceptions\ClientException
     * @throws \Bigcommerce\ORM\Client\Exceptions\ResultException
     * @throws \Bigcommerce\ORM\Exceptions\EntityException
     * @throws \Bigcommerce\ORM\Exceptions\MapperException
     */
    public function testCount()
    {
        $count = $this->repository->count();
        $this->assertEquals(2, $count);
    }

    /**
     * @covers \Bigcommerce\ORM\Repository::findAll
     * @throws \Bigcommerce\ORM\Client\Exceptions\ClientException
     * @throws \Bigcommerce\ORM\Client\Exceptions\ResultException
     * @throws \Bigcommerce\ORM\Exceptions\EntityException
     * @throws \Bigcommerce\ORM\Exceptions\MapperException
     */
    public function testFindAll(){
        $findAll = $this->repository->findAll();
        $this->assertEquals([], $findAll);
    }

    /**
     * @return object|\Prophecy\Prophecy\ProphecySubjectInterface
     */
    private function getEntityManager()
    {
        $entityManager = $this->prophet->prophesize(EntityManager::class);
        $entityManager->count(Customer::class, null)->willReturn(2);
        $entityManager->getMapper()->willReturn(new Mapper());
        $entityManager->findAll(Customer::class, null, null, false)->willReturn([]);

        return $entityManager->reveal();
    }
}